lua/diffview/vcs/adapters/hg/init.lua

```diff
@@ -1,18 +1,1051 @@
+  ---@type FileEntry[]
+  local files = {}
+  ---@type FileEntry[]
+  local conflicts = {}
+  ---@type CountDownLatch
+  local latch = CountDownLatch(3)
+  local debug_opt = {
+    context = "HgAdapter>tracked_files()",
+    func = "s_debug",
+    debug_level = 1,
+    no_stdout = true,
+  }
+
+  ---@param job Job
+  local function on_exit(job)
+    utils.handle_job(job, { debug_opt = debug_opt, fail_on_empty = false })
+    latch:count_down()
+  end
+
+  local namestat_job = Job:new({
+    command = self:bin(),
+    args = utils.vec_join(
+      self:args(),
+      "status",
+      "--modified",
+      "--added",
+      "--removed",
+      "--deleted",
+      "--template={status} {path}\n",
+      args
+    ),
+    cwd = self.ctx.toplevel,
+    on_exit = on_exit,
+  })
+  local mergestate_job = Job:new({
+    command = self:bin(),
+    args = utils.vec_join(self:args(), "debugmergestate", "-Tjson"),
+    cwd = self.ctx.toplevel,
+    on_exit = on_exit,
+  })
+  local numstat_job = Job:new({
+    command = self:bin(),
+    args = utils.vec_join(self:args(), "diff", "--stat", args),
+    cwd = self.ctx.toplevel,
+    on_exit = on_exit,
+  })
+
+  namestat_job:start()
+  mergestate_job:start()
+  numstat_job:start()
+  latch:await()
+  local out_status
+  if
+    not (#namestat_job:result() == 0 and #numstat_job:result() == 0)
+    and not (#namestat_job:result() == #numstat_job:result() - 1)
+  then
+    out_status =
+      vcs_utils.ensure_output(2, { namestat_job, numstat_job }, "HgAdapter>tracked_files()")
+  end
+
+  if out_status == JobStatus.ERROR or not (namestat_job.code == 0 and numstat_job.code == 0 and mergestate_job.code == 0) then
+    callback(utils.vec_join(namestat_job:stderr_result(), numstat_job:stderr_result(), mergestate_job:stderr_result()), nil)
+    return
+  end
+
+  local numstat_out = numstat_job:result()
+  local namestat_out = namestat_job:result()
+  local mergestate_out = mergestate_job:result()
+
+  local data = {}
+  local conflict_map = {}
+  local file_info = {}
+
+  -- Last line in numstat is a summary and should not be used
+  table.remove(numstat_out, -1)
+
+  for i, s in ipairs(namestat_out) do
+    local status = s:sub(1, 1):gsub("%s", " ")
+    local name = vim.trim(s:match("[%a%s]%s*(.*)"))
+
+    local stats = {}
+    local changes, diffstats = numstat_out[i]:match(".*|%s+(%d+)%s+([+-]+)")
+    if changes and diffstats then
+      local _, adds = diffstats:gsub("+", "")
+
+      stats = {
+        additions = tonumber(adds),
+        deletions = tonumber(changes) - tonumber(adds),
+      }
+    end
+
+    if not (kind == "staged") then
+      file_info[name] = {
+        status = status,
+        name = name,
+        oldname = name, -- TODO
+        stats = stats,
+      }
+    end
+  end
+
+  local find_key = function (t, key, value)
+    for _, v in ipairs(t) do
+      if v[key] == value then
+        return v
+      end
+    end
+  end
```

Comment:
Instead of iterating over the json data every time we're looking for a value, it would be a lot better to process the json data once, producing a struct that let's you access what you need directly.